---
title: "Arbalist analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Arbalist analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Arbalist is designed to perform scATAC-seq analysis using similar methods to [ArchR](https://www.archrproject.com/) but utilizing `MultiAssayExperiment` and the bioconductor framework instead of the ArchR framework. The backend of arbalist is based on the C++ representation in the tatami library.

# Setting up

```{r, eval=FALSE}
library(arbalist)
library(MultiAssayExperiment)
library(SummarizedExperiment)
library(SingleCellExperiment)
```

We need to create a `MultiAssayExperiment` object from the fragment files. `arbalist::saveTileMatrix` creates a count matrix for tiles across the genome and `arbalist::saveRegionMatrix` creates a count matrix for user specified genomic regions.

We are going to create some mock fragment files for this example. You should replace these with your fragment files.

```{r, eval=FALSE}
seq.lengths <- c(chrA = 6000, chrB = 5000, chrC = 3000, chrD = 2000)

fragment.files <- c();
for(i in 1:3) {
  cell.names <- paste0(LETTERS,i)
  fragment.files[i] <- tempfile(fileext = ".gz")
  mockFragmentFile(fragment.files[i], seq.lengths, num.fragments = 10, cell.names = cell.names)
}
```

First, let's create a count matrix for 500 bp genomic tiles.

```{r, eval=FALSE}
scatac.tile.matrices <- list();
tiles.grs <- NULL
for(i in seq_along(fragment.files)) {
  res <- saveTileMatrix(
    as.character(fragment.files[i]),
    output.file = tempfile(fileext=".h5"),
    output.name = 'tile_matrix',
    tile.size = 500,
    seq.lengths = seq.lengths
  )
  scatac.tile.matrices[i] <- res$counts
  
  # The genomic ranges should be the same for all the tile matrix so we are just saving one for later use
  tiles.grs <- res$tiles
}
names(scatac.tile.matrices) <- paste0('sample',seq_along(scatac.tile.matrices))
```

We use `AmalgamatedArray` to combine the dataset together so that the results for each fragment file or sample can go into its own file. 

```{r, eval=FALSE}
mat <- alabaster.matrix::AmalgamatedArray(scatac.tile.matrices, along=2)
```

Create a `SingleCellExperiment` for the tile matrix and wrap it into a `MultiAssayExperiment`.

```{r, eval=FALSE}
mat.list <- list(counts=mat)
se <-  SummarizedExperiment(mat.list)
rowRanges(se) <- tiles.grs
cell.to.sample <- unlist(sapply(names(scatac.tile.matrices),function(x) {rep(x,ncol(scatac.tile.matrices[[x]]))}))
colnames(se) <- paste0(cell.to.sample,'#',colnames(se))
colData(se)$Sample <- as.character(cell.to.sample)
tile.sce <- as(se, 'SingleCellExperiment')
el <- ExperimentList(list(
  'TileMatrix500' = tile.sce
))
maplist <- lapply(el, function(se) {
  data.frame(primary = se$Sample, colname = colnames(se), stringsAsFactors = FALSE)
})
sampMap <- listToMap(maplist)
mae <- MultiAssayExperiment(el, sampleMap = sampMap, colData = DataFrame(row.names=unique(sampMap$primary)))
```

Finally, we add the fragment file names to the `MultiAssayExperiment` `colData` for use by arbalist functions.

```{r,eval=FALSE}
colData(mae)$fragment_file <- fragment.files
```

# Quality control

## Fragment size distribution

Plotting fragment size distributions
```{r fragplots, eval=FALSE}
sample.names <- rownames(colData(mae))
for(i in sample.names) {
  plotFragmentSizeDistribution(mae, i)
}
```

## TSS enrichment

The TSS enrichment score evaluates accessibility at the transcription start site compared to flanking regions in the genome. The start coordinate of each gene is used as the transcription start site. For multiome (scATAC+RNA) datasets, a gene matrix can also be added to the `MultiAssayExperiment` as the GeneExpressionMatrix and arbalist will default to extracting gene ranges from the GeneExpressionMatrix. For datasets without a GeneExpressionMatrix, gene genomic ranges need to be specified.

```{r tss, eval=FALSE}
library(GenomicRanges)
gene.grs <- GRanges(c("chrA:500-1000", "chrA:1200-1400", "chrA:1500-2700", "chrA:4500-5000","chrB:1000-2000", "chrB:3500-4200", "chrC:1500-2000", "chrC:2500-3000", "chrD:500-1000"))

colData(mae)
mae <- addTSSEnrichmentScores(mae, gene.grs = gene.grs)
colData(mae)
```
A TSSEnrichment column was added to the MultiAssayExperiment colData.

## Sequencing doublets

We calculate doublet scores and filter cells based on the doublet enrichment score.

```{r doublet}
mae <- addDoubletScores(mae)
summary(colData(mae[['TileMatrix500']])$DoubletScore)
summary(colData(mae[['TileMatrix500']])$DoubletEnrichment)
mae[['TileMatrix500']] <- mae[['TileMatrix500']][,colData(mae[['TileMatrix500']])$DoubletEnrichment >= 1]
```

# Reduced dimensions using iterative LSI

We use iterative LSI for dimensionality reduction of the scATAC-seq.

```{r lsi, eval=FALSE}
mae <- addIterativeLSI(mae, embedding.name = 'iterativeLSI')
```

# Cell clusters

We calculate clusters of cells using the addClusters function, which is a wrapper on Seurat functions.

```{r clusters, eval=FALSE}
mae <- addClusters(mae, clusters.colname = 'Clusters')
plotClusterConfusionMap(mae)
```

# Embeddings

We calculate UMAP and tSNE embeddings of the reduced dimension features. Embeddings are useful for visualization.

```{r embeddings, eval=FALSE}
library(scater)

# Add UMAP to SingleCellExperiment
mae[['TileMatrix500']] <- runUMAP(mae[['TileMatrix500']], dimred = 'iterativeLSI', name = 'iterativeLSI_UMAP')

# UMAP colored by Sample
scater::plotReducedDim(mae[['TileMatrix500']], dimred = 'iterativeLSI_UMAP', colour_by = 'Sample')

# UMAP colored by Clusters
scater::plotReducedDim(mae[['TileMatrix500']], dimred = 'iterativeLSI_UMAP', colour_by = 'Clusters')

# Add TSNE to SingleCellExperiment
mae[['TileMatrix500']] <- runTSNE(mae[['TileMatrix500']], dimred = 'iterativeLSI', name = 'iterativeLSI_TSNE')

# TSNE colored by Sample
scater::plotReducedDim(mae[['TileMatrix500']], dimred = 'iterativeLSI_TSNE', colour_by = 'Sample')

# TSNE colored by Clusters
scater::plotReducedDim(mae[['TileMatrix500']], dimred = 'iterativeLSI_TSNE', colour_by = 'Clusters')
```

# Pseudobulk replicates

```{r pseudobulk, eval=FALSE}
mae <- addGroupCoverages(mae)
```

`arbalist::addGroupCoverages` performs 3 tasks. First, a pseudobulk `SummarizedExperment` is added to the `MultiAssayExperiment`. The `skip.se.creation` argument to `addGroupCoverages` allows you to skip creating the `SummarizedExperiment`.

```{r, eval=FALSE}
mae
```

Second, the cell composition of the pseudobulk replicates is added to the `SingleCellExperiment` `colData`. 

```{r, eval=FALSE}
colData(mae[['TileMatrix500']])
```

Third, pseudobulk coverage files containing the fragments coordinates are written to file and file names are specified in the pseudobulk \linkS4class{SummarizedExperiment} colData. The `skip.coverage.file.creation` argument to `addGroupCoverages` allows you to skip creating the coverage files. However, these are required to run `addPeakMatrix`.

```{r, eval=FALSE}
colData(mae[['TileMatrix500_pseudobulk']])
```

# Peak calling

Peak calling is performed with MACSr. MACS peak calling requires the effective genome size is specified or defaults to human genome. MACS recommends 2.7e9 for the human genome and 1.87e9 for the mouse genome. `maw.scatac::getEffectiveGenomeSize` uses the `MultiAssayExperiment` metadata to look up the genome version and then returns the recommended effective genome size.

```{r peaks, eval=FALSE}
mae <- addPeakMatrix(mae, genome.size = maw.scatac::getEffectiveGenomeSize(mae[[1]]), output.dir = '.')
mae
```
